chore(improve-importer): init

pongstr opened the pull request:
Co-authored-by: Micha Mazaheri <micha@mazaheri.me>

- updated: utilize @apidevtools/swagger-parser to validate and parse OpenAPI documents
- updated: utilize [openapi-types](https://github.com/kogosoftwarellc/open-api/tree/master/packages/openapi-types#readme) for typings
- chore: upgraded webpack, babel setup and jest setup
- chore: updated extension config references
- chore: simplify code coherence, reduce code complexity

**TODO**
- [ ] `utils/paw-url.ts` improve code coherence, utilize `URL` api
- [ ] `utils/json-schema-parser.ts` - improve code, ensure compliance with [openapi/json-schema](https://tools.ietf.org/html/draft-wright-json-schema-00#section-4.2) specs
- [ ] `utils/console.ts` - ensure to cover cyclic objects+references

pongstr committed:
🗄️ chore(openapi-updates): init

pongstr committed:
🗄️ chore(openapi-updates): format paw typings

pongstr committed:
🗄️ chore(openapi-updates): format paw typings

pongstr committed:
🗄️ chore(openapi-updates): use openapi-types lib

pongstr committed:
🗄️ chore(openapi-updates): use openapi-types lib

pongstr committed:
🗄️ chore(openapi-updates): init utils revision

pongstr committed:
🧽 chore(openapi-updates): better console abstraction

pongstr committed:
chore(openapi-updates): configs + script cleanup

pongstr committed:
🗄️ chore(openapi-updates): linting+formatting

pongstr committed:
🗄️ chore(openapi-updates): change target to web

changing the target from 'node-webkit' to 'web' allows us to utilize
node.js api(s) that are polyfilled by webpack/babel, which should work
with webkit since webkit still complies with web standards.

pongstr committed:
chore(openapi-updates): setImmediate isnt in context

pongstr committed:
rm refactored modules

pongstr committed:
init for discussion

mittsh committed:
Fix loading

mittsh committed:
Remove needless new Promise() creation

mittsh committed:
Fix location

mittsh committed:
Add test request and some comments

pongstr committed:
Merge pull request #7 from luckymarmot/async-swagger-parser-mittsh

Fix async calls

pongstr committed:
updates

pongstr committed:
updates

pongstr committed:
- update: remove logs

pongstr committed:
updated: fix env var attachments

pongstr committed:
chore (improve-importer): update typings+config import

mittsh commented on src/lib/converter.ts:
```diff
@@ -0,0 +1,434 @@
+      variable.required = required || false
+
+      if (param.in === 'path') {
+        envManager.setEnvironmentVariableValue(
+          param.name,
+          parameterExampleValue,
+        )
```
This actually should be a "request variable" instead of an environment variable.

![image](https://user-images.githubusercontent.com/1319906/115695063-1fc81800-a36a-11eb-8eb3-150b1716f0c9.png)

You actually have the right snippets elsewhere.

To create the variable, already done on line 311 above:

```ts
    const variable = request.addVariable(
        name,
        parameterExampleValue,
        description || '',
      )
```

To use the variable, I believe you have the right snippet in `convertEnvString()`:

```ts
    let requestVariable = request.getVariableByName(match[1])
      if (requestVariable && requestVariable.id) {
        components.push(
          new DynamicValue('com.luckymarmot.RequestVariableDynamicValue', {
            variableUUID: requestVariable.id,
          }),
        )
      }
```

pongstr replied:
@mittsh I have no clue how to move from this tbh.

mittsh commented on src/lib/converter.ts:
```diff
@@ -0,0 +1,434 @@
+    const { title } = document.info
+
+    const request = this.context.createRequest(
+      item.summary || item.path,
```
Perhaps you could even use `operationId` here for a nicer name:

```suggestion
      item.summary || item.operationId || item.path,
```

mittsh commented on src/lib/converter.ts:
```diff
@@ -0,0 +1,434 @@
+    }
+
+    if (item.responses) {
+      request.setHeader('Accepts', item.responses[0])
```
```suggestion
      request.setHeader('Accept', item.responses[0])
```

`Accept` is the right spelling :) 
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

mittsh commented on src/lib/converter.ts:
```diff
@@ -0,0 +1,434 @@
+    const schema = requestBody.content[mediaTypes[0]]
+      .schema as OpenAPIV3.SchemaObject
+
+    request.body = JSON.stringify(jsonSchemaParser(schema, {}) as any, null, 2)
```
If you're sure it's a JSON body, you could even use:

```suggestion
    request.jsonBody = jsonSchemaParser(schema, {}) as any
```

Note: `jsonBody` should be an object, so no need for `JSON.stringify()`

pongstr replied:
It's definite the body will always be in [JSON|JSONSchema object](https://swagger.io/specification/#schema-object) (if provided in the OpenAPI v3 document, either as inline or component reference).

mittsh commented on src/lib/converter.ts:
```diff
@@ -0,0 +1,434 @@
+      const variable = request.addVariable(
+        name,
+        parameterExampleValue,
+        description || '',
```
Suggesting this as I get descriptions with a trailing new line or spaces

```suggestion
        (description || '').trim(),
```

mittsh commented on examples/openapi-official-examples/petstore.yaml:
```diff
@@ -1,111 +1,826 @@
-openapi: "3.0.0"
+---
```
I'm having trouble importing the following example files:

- `api-with-examples.yaml`
- `uspto.yaml`

Not sure why, it does not import any item:

![image](https://user-images.githubusercontent.com/1319906/115697650-6585e000-a36c-11eb-8acb-7255391e6a8f.png)

pongstr replied:
added the fix for this bug, the requests were being appended to non-existing group names. should be working now.

mittsh commented on src/lib/converter.ts:
```diff
@@ -0,0 +1,434 @@
+        description || '',
+      )
+
+      variable.schema = schema as any
```
I suggest removing the `type` and `format` keys from the schema.

```suggestion
      const schema = {
        ...schema as any
      }
      delete schema.type
      delete schema.format
      variable.schema = schema
```

These two keys make Paw consider the schema as invalid. Removing those keys we have this nice popover window:

![image](https://user-images.githubusercontent.com/1319906/115698910-cd88f600-a36d-11eb-8441-d9c083ad4e99.png)

mittsh reviewed (commented):
This works really well! Tested with all the examples that we have.

Also, the grouping technique for requests is excellent. Looks beautiful 👌

Some comments below.

mittsh commented on src/lib/converter.ts:
```diff
@@ -0,0 +1,434 @@
+        access_token_uri: authFlow.tokenUrl,
+        redirect_uri: '',
+        scope: `'${item.value}'`,
+        state: undefined,
+        token: undefined,
+        token_prefix: undefined,
+        grant_type: snakeCase(grantType),
```
I forgot to add: unfortunately `grant_type` has to be always `'authorization_code'`. It's something we can improve later in Paw.

Also, I'd recommend avoiding the `state`, `token`, and `token_prefix` parameters altogether (the TypeScript definition is wrong, we are allowed to omit these params).

```suggestion
        grant_type: 'authorization_code',
```

mittsh commented on src/lib/converter.ts:
```diff
@@ -0,0 +1,434 @@
+      request.oauth2 = {
+        client_id: '',
+        client_secret: '',
+        authorization_uri: authFlow.authorizationUrl,
+        access_token_uri: authFlow.tokenUrl,
```
Finally found why it didn't work, `authFlow.tokenUrl` was `undefined`, therefore `access_token_uri` was also `undefined`, but actually a required param for Paw.

This should solve the problem:

```suggestion
        authorization_uri: authFlow.authorizationUrl || '',
        access_token_uri: authFlow.tokenUrl || '',
```

pongstr committed:
chore(pr-revisions): update

pongstr committed:
chore(pr-revisions): empty string groups

mittsh committed:
Fix use of request variables

pongstr committed:
chore(improve-importer): init

Co-authored-by: Micha Mazaheri <micha@mazaheri.me>

- updated: utilize @apidevtools/swagger-parser to validate and parse OpenAPI documents
- updated: utilize [openapi-types](https://github.com/kogosoftwarellc/open-api/tree/master/packages/openapi-types#readme) for typings
- chore: upgraded webpack, babel setup and jest setup
- chore: updated extension config references
- chore: simplify code coherence, reduce code complexity

**TODO**
- [ ] `utils/paw-url.ts` improve code coherence, utilize `URL` api
- [ ] `utils/json-schema-parser.ts` - improve code, ensure compliance with [openapi/json-schema](https://tools.ietf.org/html/draft-wright-json-schema-00#section-4.2) specs
- [ ] `utils/console.ts` - ensure to cover cyclic objects+references
